### PR TITLE
Introduce K8sClient abstraction for easier testing

### PR DESCRIPTION
**Pull request by swoehrl-mw**
This PR introduces a new interface `K8sClient` as an abstraction over the existing `client.Client` and `ResourceReconciler`. The goals are to reduce some boilerplate code and make it easier to unittest the code with mocks instead of using the slow envtest approach everywhere.

Changes:
* New interface `K8sClient` with `K8sClientImpl` as default implementation that wraps all kubernetes client access
* Introduce [mockery](https://vektra.github.io/mockery/) as helper library to generate mocks
* Refactor all tests in `pkg/reconcilers` to use a `MockK8sClient`

With these changes it should be easier to write unittests for the reconcilers in the future. Also the runtime of the tests should be significantly reduced.

### CODE REVIEW

**Comment by saketmht** [comment]
@swoehrl-mw What is pending in this PR?

**Comment by swoehrl-mw** [comment]
> What is pending in this PR?

@saketmht 
It's waiting for a review and approval. After that I will fix the conflicts before merging (I'm not gonna fix them beforehand as there would be new ones with each other PR merged).

File: opensearch-operator/pkg/reconcilers/role_test.go [hunk 1]
```diff
@@ -61,25 +62,7 @@ var _ = Describe("roles reconciler", func() {
 				},
 			},
 		}
-
-		// Sleep for cache to start
-		time.Sleep(time.Second)
 		// Set up prereq-objects
-		ns = &corev1.Namespace{
```
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/role_test.go:68]: Can you please explain why we deleted that step ?

File: opensearch-operator/pkg/reconcilers/scaler.go [hunk 2]
```diff
@@ -108,8 +107,9 @@ func (r *ScalerReconciler) reconcileNodePool(nodePool *opsterv1.NodePool) (bool,
 	// Also the status is set to 'Running' if it fails to exclude node for some reason
 	if !found || currentStatus.Status == "Running" {
 		// Change the status to running, to indicate that a scaling operation for this nodePool has started
-		r.instance.Status.ComponentsStatus = helpers.Replace(currentStatus, componentStatus, r.instance.Status.ComponentsStatus)
-		err := r.Status().Update(r.ctx, r.instance)
+		err := r.client.UpdateOpenSearchClusterStatus(client.ObjectKeyFromObject(r.instance), func(instance *opsterv1.OpenSearchCluster) {
```
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/scaler.go:110]: is the .UpdateOpenSearchClusterStatus is client func or our ?

File: opensearch-operator/pkg/reconcilers/suite_test.go [hunk 3]
```diff
@@ -72,94 +46,20 @@ func TestAPIs(t *testing.T) {
 var _ = BeforeSuite(func() {
 
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
-	ports, err := freeport.GetFreePorts(2)
-	Expect(err).NotTo(HaveOccurred())
-
-	serviceMonitorCRD := apiextensionsv1.CustomResourceDefinition{
```
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/suite_test.go:78]: did we moved the build of serviceMonitorCRD to somewhere else ?

**Review by idanl21 — APPROVED** [review]

**Pull request merged**